server/src/main/java/org/opensearch/common/util/MovingAverage.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.util;
+
+/**
+ * MovingAverage is used to calculate the moving average of last 'n' observations.
+ */
+public class MovingAverage {
+    private final int windowSize;
+    private final long[] observations;
+
+    private long count = 0;
+    private long sum = 0;
+    private double average = 0;
+
+    public MovingAverage(int windowSize) {
+        if (windowSize <= 0) {
+            throw new IllegalArgumentException("window size must be greater than zero");
+        }
+
+        this.windowSize = windowSize;
+        this.observations = new long[windowSize];
+    }
+
+    /**
+     * Records a new observation and evicts the n-th last observation.
+     */
+    public synchronized double record(long value) {
+        long delta = value - observations[(int) (count % observations.length)];
+        observations[(int) (count % observations.length)] = value;
+
+        count++;
+        sum += delta;
+        average = (double) sum / Math.min(count, observations.length);
+        return average;
+    }
```

Comment:
Can we use a queue and see if we can avoid the synchronized block by using CAS if possible

Reply:
I actually plan to benchmark CAS vs synchronized approaches before committing to either one of them, especially since the operations are pretty simple and quick to execute.

If there are major gains with CAS then it makes sense to go with it. Otherwise, I would prefer keeping it simple and readable.

Reply:
I did not observe any major gains with queue + CAS approach as followed in indexing back-pressure ([here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/ShardIndexingPressureTracker.java#L226-L227)). With higher thread contention, it performed slightly poorly.

```
# JMH version: 1.35
# VM version: JDK 17.0.3, OpenJDK 64-Bit Server VM, 17.0.3+7

Benchmark (1 thread - no contention)           Mode  Cnt   Score   Error  Units
MovingAverageBenchmark.timeMovingAverage       avgt    5  25.669 ± 1.884  ns/op
MovingAverageBenchmark.timeMovingAverageQueue  avgt    5  25.213 ± 0.383  ns/op

Benchmark                                      Mode  Cnt    Score   Error  Units
MovingAverageBenchmark.timeMovingAverage       avgt    5  217.714 ± 6.676  ns/op
MovingAverageBenchmark.timeMovingAverageQueue  avgt    5  223.088 ± 3.651  ns/op

Benchmark (16 threads - high contention)       Mode  Cnt    Score   Error   Units
MovingAverageBenchmark.timeMovingAverage       avgt    5  785.830 ± 13.446  ns/op
MovingAverageBenchmark.timeMovingAverageQueue  avgt    5  792.442 ± 64.234  ns/op
```

It is also worth noting that the current implementation of moving average in shard indexing back-pressure has subtle race-condition bugs. Two or more concurrent threads could reach this [point](https://github.com/opensearch-project/OpenSearch/blob/4b08d386e7e567071365029bd1d71b3bb0388cc7/server/src/main/java/org/opensearch/index/ShardIndexingPressure.java#L282) and remove excessive elements from the queue leading to incorrect results, or even NPE in the worst case. Another problem is lack of causal ordering – an older thread may overwrite a new thread's average at this [point](https://github.com/opensearch-project/OpenSearch/blob/4b08d386e7e567071365029bd1d71b3bb0388cc7/server/src/main/java/org/opensearch/index/ShardIndexingPressure.java#L293).